typing/typedtree.ml

```diff
@@ -651,10 +651,16 @@ and 'a class_infos =
    }
 
 type implementation = {
-  structure: structure;
-  coercion: module_coercion;
-  signature: Types.signature;
-  shape: Shape.t;
+  impl_structure: structure;
+  impl_coercion: module_coercion;
+  impl_signature: Types.signature;
+  impl_shape: Shape.t;
+  impl_loc: Location.t
```

Comment:
I am not in favor of introducing a field prefix where none was used (here and on `interface` below). OCaml supports type-based disambiguation of fields and constructors now, so we don't need to impose unicity of field names. A part of the compiler types were written before this feature was added, but it is not clear that we need to stick to this visually-noisy style now. I would support the existing code as it is, which means keeping the prefixes on prefixed field names, but also keeping unprefixed names as they are.

Reply:
Should loc also be unprefixed? It causes ambiguity with Location.loc that the compiler can't resolve, and would cause a lot of changes in the form of type annotations in both the compiler code and other code. It also means that the interface type would have a subset of the fields of the implementation type (signature and loc) which would also cause ambiguity. Alternatively, we can give it another name there.

Reply:
Good question... I just gave `location` a try at https://github.com/gasche/ocaml/commits/add-interface-and-implementation-types and it appears to work out. (One has to annotate record-of-functions arguments that also have a `location` method, but I think this is a reasonable change.)

I get your point that the prefixing you did was in fact the more minimal change. (It is a bit ugly where it is used, but it does not affect the rest of the code.) Do you have a preference between `location` and an `impl_`/`intf_` suffix?

(Note: I did not notice any specific ambiguity between `implementation` and `interface`; either it is so rare that I don't remember it, or the typing gives enough information to disambiguate always.)

Reply:
Thanks for the attempt! My slight preference is for a prefixed version to be more consistent with the rest of the AST, and to avoid confusing compile messages I encountered while making changes (a type was improperly inferred at a function, and the inference propagates up the call stack and manifests itself in a call site where it's hard to tell where the inference pointed at the wrong type). However, my preference is only slight, so I'll defer to your opinion since you're more familiar with the code base and the desired coding style.
Side note: has anyone ever gotten Merlin/LSP working within the compiler code base?

Reply:
Let's ask @Octachron for a preference. (Anyone is of course free to chime in, but we two got you into this mess so we are morally encouraged to do it.)

>  Side note: has anyone ever gotten Merlin/LSP working within the compiler code base?

Sure, it is a pain but it works well: https://github.com/ocaml/ocaml/blob/3d4eb0663a1c474bd97bf17c88450ab7fac35141/HACKING.adoc#using-merlin

Reply:
Thanks! And no worries at all!

Reply:
If the issue is with the multiple uses of `loc`, I would advocate to be more precise on this single field and use `impl_source_loc` (resp. `intf_source_loc`)  to explicit that the field provides us location information on the source file.

Reply:
@Octachron for me the question is whether we prefer to change the `Typedtree.implementation` type to use an `impl_` prefix where none was previously used . If the answer is 'no prefix', then indeed a secondary question is how to name the location field (and you propose `source_loc`).

Note: do we actually need a full location or is a source filename enough?

Reply:
I would rather don't add a prefix in a PR with a distinct objective.
And you are right, that we could keep just the filename rather a full `Location.t`. The [current error message](https://github.com/ocaml/ocaml/blob/trunk/typing/includemod.ml#L980) don't use more information, and it is doubtful if the size of the file will ever be useful.

Reply:
So I guess `file` or `fname` would introduce less naming ambiguities than `loc` :-)